### ISSUE TITLE
Geo: Proposal for refactoring out ShapeBuilder hierarchy 

### ISSUE THREAD

**Issue by imotov** [elastic/elasticsearch#40908]
At the moment  shapeBuilders perform much more functionality than just shape building. We have serialization, parsing logic plus logic necessary for handling dateline transformation and conversion of shapes for lucene indexing in both libs/geo and S4J formats. As we are removing support for the S4J shapes, I would like to propose refactoring ShapeBuilder set of classes. 

In #36477 we introduced a hierarchy of geo classes that can replace the hierarchy build with ShapeBuilder as a root. This hierarchy has pretty much all functionality except GeoJson parser and logic that is needed to convert shapes into the format consumable by Lucene, if we could add the GeoJson parser to libs/geo, and extract the logic we can remove the entire ShapeBuilder hierarchy and significantly simplify the geo processing pipeline structure. I think it will also simplify switching to different reference system, since it will allows us to have the reference system-specific logic separate from the geo hierarchy. 

It will also remove the issue that we are facing in geosql at the moment. The only way to retrieve a shape in geo at the moment is through the`ShapeParser.parse(....).buildGeometry()` path, which converts some shapes into somewhat equivalent set of shapes for lucene consumption, among other things it changes the order of coordinates of polygons, splits polygons on dateline, simplifies collections, etc. This can be really confusing for geosql users, since they will get  shapes different from what they put into elasticsearch.

To summarize the following actions will be required:

- [X] Add GeoJson Parser to libs/geo #41575
- [x] Extract dateline decomposition and other Lucene-specific transformation into a separate set of classes. #44187
- [ ] Remove Shape Builder together with S4J logic and replace it with libs/geo and extracted transformation classes.

**Comment by talevy** [comment]
+1

**Comment by iverase** [comment]
@imotov in https://github.com/elastic/elasticsearch/pull/77856 we moved all ShapeBuilder classes to a module together with the Spatial4j/JTS dependency, are we ok to close this issue?

**Comment by iverase** [comment]
closing in favour of https://github.com/elastic/elasticsearch/issues/96097
